snuba/query/processors/uuid_column_processor.py

```diff
@@ -0,0 +1,112 @@
+from typing import Set
+import uuid
+
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.query.conditions import (
+    binary_condition,
+    is_condition,
+    is_in_condition,
+    is_not_in_condition,
+)
+from snuba.query.logical import Query
+from snuba.query.processors import QueryProcessor
+from snuba.request.request_settings import RequestSettings
+
+
+class UUIDColumnProcessor(QueryProcessor):
+    """
+    If a condition is being performed on a column that stores UUIDs (as defined in the constructor)
+    then change the condition to use a proper UUID instead of a string.
+    """
+
+    def __init__(self, uuid_columns: Set[str]) -> None:
+        self.__uuid_columns = uuid_columns
+
+    def is_uuid_column(self, exp: Expression) -> bool:
+        return isinstance(exp, Column) and exp.column_name in self.__uuid_columns
+
+    def is_formatted_uuid(self, exp: Expression) -> bool:
+        return (
+            isinstance(exp, FunctionCall)
+            and exp.function_name == "replaceAll"
+            and isinstance(exp.parameters[0], FunctionCall)
+            and exp.parameters[0].function_name == "toString"
+            and isinstance(exp.parameters[0].parameters[0], Column)
+            and exp.parameters[0].parameters[0].column_name in self.__uuid_columns
+        )
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        def process_condition(exp: Expression) -> Expression:
+            if is_condition(exp):
+                assert isinstance(exp, FunctionCall)
+                if len(exp.parameters) > 1:
+                    # This doesn't seem great but here it is: There is a separate processor that converts
+                    # all cases of UUID columns to do a string comparison. We want to to undo that work
+                    # for conditions only, so we need to handle the common case (no formatting) as well
+                    # as the case that the other processor has formatted the column already.
+
+                    # Check if we have a uuid column
+                    if any(
+                        self.is_uuid_column(param) or self.is_formatted_uuid(param)
+                        for param in exp.parameters
+                    ):
```

Comment:
I think there is an issue with this approach, you are transforming any expression that is a condition in the query, which will cover all the cases in the conditions clause (which is good) and potentially conditions used elsewhere (I hope that does not happen), which is harmless though not incredibly useful but it would miss a `GROUP BY replaceAll(toString(....)`, which can be useful.

I would suggest you just selectively apply this transformation to the condition, prewhere and group by clauses.

Reply:
```
condition = query.get_condition_from_ast()
for e in condition:  # Expressions are iterables that actually traverse the graph in. postfix order.
      do_your_transformation
```
Same for group by and prewhere